log4j2/src/main/java/com/google/common/flogger/backend/log4j2/Log4j2LogEventUtil.java

```diff
@@ -52,7 +57,15 @@ private Log4j2LogEventUtil() {}
   static LogEvent toLog4jLogEvent(String loggerName, LogData logData) {
     MetadataProcessor metadata =
         MetadataProcessor.forScopeAndLogSite(Platform.getInjectedMetadata(), logData.getMetadata());
-    String message = SimpleMessageFormatter.getDefaultFormatter().format(logData, metadata);
+    // TODO: Check if the user of the log4j2 backend requires more flexibility when deciding which
+    //       formatter to use. I'd argue that the BaseMessageFormatter is sufficient, whenever a log4j2.xml
+    //       file is present. However, users might be fine with not using log4j2 specifics when using flogger
+    //       and then it is sufficient to use the default formatter.
+    // For the moment, I'd argue we want to pass the context data to log4j2 and make use of a log4j2
+    // condiguration file. Hence the line below ist commented out.
+    // String message = SimpleMessageFormatter.getDefaultFormatter().format(logData, metadata);
+    String message = BaseMessageFormatter.appendFormattedMessage(logData, new StringBuilder()).toString();
```

Comment:
@cgdecker thank you very much for investing your time. I am very sorry for the delay. 

> I guess the idea is that this version doesn't include metadata in the message, instead relying on the fact that said metadata is being passed on to log4j2 which can choose how to format that itself?

This is correct. From my point of view, the flexibility provided by Log4j2 Appenders and Log4j2 Layouts is one of the strengths of Log4j2. Maybe this kind of flexibility is not required by flogger, however, I'd argue that users, who choose to use the log4j2 backend implementation may also be interested in configuring appenders and layouts. Or at least they may want to have the option to use them. 

Since the message constructed by flogger includes the metadata, it poses a problem when a user wants to log4j2 layout to format a LogEvent.

> That sounds reasonable to me, but I don't personally have any knowledge of what users would expect or whether this might cause any issues for existing users.

I also feel a bit uncomfortable using the `BaseMessageFormatter`, as there might be users who want to use the default behavior. In case you want to add the capability of leaving the formatting of a `LogEvent` to the log4j2 backend, I believe, it might be reasonable to add this as an experimental feature that can be enabled and is disabled by default.

How can we implement this? I am not 100% sure what is the best way. There might be at least three ways to implement it:
_1) Enable it when a log4j2 configuration file is located_
[Log4j2 Configuration](https://logging.apache.org/log4j/log4j-2.2/manual/configuration.html) states the steps taken in order to automatically configure Log4j2 during its initialization. Flogger's Log4j2 implementation might check if a configuration file is loaded.

_2) Introduce a Flogger Log4j2 backend specific flag_
The easiest solution might be introduce another system variable being used a flag. However, this may produce an unwanted behavior in case the flag is not set and a log4j2 configuration file is loaded.

_3) Add a second log4j2 backend_
Keeps this logic sompletely seperate from the original implementation and might be useful in order to evaluate this feature. This approach will increase maintenance efforts and may cause confusion as there are two log4j2 backends.

 What do you think?

Reply:
I wouldn't get super worried about mimicking the current metadata formatting in `SimpleMessageFormatter`. That class was never intended to be some foundational thing all the other backends would use, but since it was public and easy, they did. The metadata formatting it does was largely something I made up in a hurry with the hope that someone would come along and just do a better job later. Internally in Google we already use a richer formatter which can do things like reposition specific metadata within the log statement.

Splitting `BaseMessageFormatter` out and having things like `MetadataProcessor` is a deliberate attempt to allow different backends to process metadata in a more structured way that's in keeping with the style expected by the users of each different logging library. So I think the main things I care about here are:

1) Metadata should not be dropped by default.
2) Metadata for the log4j2 backend is handled in a way that's as familiar as possible to log4j2 users.
3) Existing users who get the "all in one" formatting aren't broken by this PR.

Now since we *don't* expect log4j2 users to have already written formatter configs to handle specific flogger data specially, we would ideally want some default formatting for any unrecognized metadata (to satisfy point 1). But this is where there has to be some explicit handshaking between the two systems.

Perhaps the right long term approach is to have a Flogger aware log4j2 formatter which can perhaps be installed as a default if nothing else is present, and understands certain default expectations around metadata formatting. However this idea is largely me guessing since I really just don't have the experience with log4j (1 or 2) to know if this makes sense, so please feel free to tell me I'm wrong here.

I'm just going from the idea of someone with a code base which mixes "normal" log4j2 and Flogger log statements and just want things to work "out the box" without dropping data, and giving log statements which are formatted consistently between Flogger and "normal" log4j2 use.

In short I don't know yet what the best thing to do is, and I'm happy to take your advice. We don't need to make this perfect straight away and can start with the simplest sensible approach that's strictly better than what we have now.

Reply:
I think there is a decent chance to satisfy all three points as there is a way to detect if a log4j configuration file is present The [Log4j2 Documentation section Automatic Configuration](https://logging.apache.org/log4j/2.x/manual/configuration.html) states that 

> If no configuration file could be located the DefaultConfiguration will be used. This will cause logging output to go to the console.

The default configuration ignores the context data and does not format. 

We may use the `BaseMessageFormatter` only if log4j2 configuration file is present, which can be checked by testing if the `org.apache.logging.log4j.core.config.Configuration` is not an instance of `org.apache.logging.log4j.core.config.DefaultConfiguration`, otherwise we use the `SimpleMessageFormatter`.

Below is a code snippet illustrating it:
```java
LoggerContext ctx = (LoggerContext) LogManager.getContext(false);
Configuration config = ctx.getConfiguration();
String message;

if (config instanceof DefaultConfiguration) {
        message = SimpleMessageFormatter.getDefaultFormatter().format(logData, metadata);
} else {
        message = BaseMessageFormatter.appendFormattedMessage(logData, new StringBuilder()).toString();
}
```

The log4j2 configuration also mentions 4 ways of how log4j2 can be configured.
>  Configuration of Log4j 2 can be accomplished in 1 of 4 ways:
    1. Through a configuration file written in XML, JSON, YAML, or properties format.
    2. Programmatically, by creating a ConfigurationFactory and Configuration implementation.
    3. Programmatically, by calling the APIs exposed in the Configuration interface to add components to the default configuration.
    4. Programmatically, by calling methods on the internal Logger class.

The proposed solution only covers the first point. I do not think you can detect if log4j2 was configured programmatically. 

> Perhaps the right long term approach is to have a Flogger aware log4j2 formatter which can perhaps be installed as a default if nothing else is present, and understands certain default expectations around metadata formatting. 

Maybe the proposed solution is a step in this direction.

Reply:
That sounds very promising. If you can do that in this PR and it would keep existing unconfigured users seeing the metadata, that sounds like a win. Having an override to force one or other behaviour might end up being necessary, but I think right now this looks to be good enough. Please add a bunch of explanatory comments for this (and if needed, guard that cast in some way, or at least comment why it's safe).

Reply:
Also, we can always add a way to inject something different to the SimpleMessageFormatter if needed later. Hopefully the Log4j formatters are rich enough we won't need to confuse things by adding another layer of formatters and configuration though.

Reply:
Unfortunately, the cast is needed because `LoggerContext.getContext` returns `org.apache.logging.log4j.spi.LoggerContext` and we need `org.apache.logging.log4j.core.LoggerContext`. I am afraid accessing the Log4j2 configuration through the internal API is the only way as of now. 

Regarding the mid/long-term vision, using a ConfigurationBuilder with a custom ConfigurationFactory might be the right way as explained [here](http://logging.apache.org/log4j/2.x/manual/customconfig.html#Hybrid).

Reply:
Thanks for explaining. Legacy APIs are fun!!

If the cast is 100% safe, just add a comment explaining it (ideally with a URL linking to official docs saying this).
Otherwise catch the class cast exception and default to the simple formatter.

Reply:
Hey, I figured out that you can read the `LoggerContext` in a slightly different way since log4j2 2.14.1, see: https://github.com/apache/logging-log4j2/commit/b0316f5302329c93d2430b03011629891a3c96ba

With this the cast is not needed.